feat(artifact): add S3-compatible artifact storage service

grignolalouis opened the pull request:
Add support for AWS S3 and S3-compatible object storage services
  as artifact storage backend.

  Features:
  - Functional options pattern for flexible configuration
  - Support for IAM roles, environment variables, and ~/.aws/credentials
  - Session and user namespace support
  - Automatic versioning
  - aws sdk for go v2 version compatible with go 1.21
 
  Changes:
  - modify  en/artifact.md and zh/artifact.md

  Review:
  - review zh/artifact.md since i do not speak chinese.

  RELEASE NOTES: Add S3-compatible artifact storage service 
  
Fixes #761

grignolalouis commented:
I have read the CLA Document and I hereby sign the CLA

liuzengh commented on docs/mkdocs/zh/artifact.md:
````diff
@@ -95,6 +95,47 @@ import "trpc.group/trpc-go/trpc-agent-go/artifact/cos"
+#### 使用方法
+
+```go
+// 创建服务 - 凭证从环境变量、~/.aws/credentials 文件或 IAM 角色（在 AWS 基础设施上）读取
````
文档应明确凭证在多种方法可用时的优先级。

  注释提及凭证从“环境变量、~/.aws/credentials 文件或 IAM
  角色”中读取。明确这些方法的优先顺序将有助于避免混淆，因为 AWS SDK 通常有明确的优先级。

liuzengh commented on artifact/s3/client.go:
```diff
@@ -0,0 +1,309 @@
+
+	// HeadObject retrieves object metadata without downloading the object.
+	HeadObject(ctx context.Context, key string) (*ObjectInfo, error)
+}
```
Do not export interfaces that the users of the package do not need.  
https://google.github.io/styleguide/go/decisions#interfaces
make the Client interface unexported if you want to use it at test

grignolalouis replied:
done !

liuzengh commented on artifact/s3/errors.go:
```diff
@@ -0,0 +1,49 @@
+	ErrBucketNotFound = errors.New("s3: bucket not found")
+)
+
+// IsNotFound reports whether the error indicates that an object was not found.
```
Consider whether the IsX helpers are necessary

errors.Is(err, ErrX) is already concise. Helpers are fine if they improve readability in call sites, but if they're only thin wrappers with no added semantics they might be unnecessary. Keep them only if used widely.

grignolalouis commented on artifact/s3/client.go:
```diff
@@ -0,0 +1,276 @@
+	}
+
+	// Check for NoSuchKey error
+	var noSuchKey *types.NoSuchKey
```
thanks for your feedback !

grignolalouis replied:
fixed

grignolalouis commented:
- Removed DeleteObject - DeleteObjects handles single and batch deletes, no need for both
  - Removed DisableSSL option - determined by endpoint URL (http:// vs https://)
  - Removed Timeout / WithTimeout - Caller can manage timeout via context.WithTimeout, keeps code simple should we keep this option ?
  -  bucket string stored now in the underlying client
  - Simplified objectInfo struct to []string we only need the keys

Error handling

  - Removed IsX helper functions - errors.Is(err, ErrX) is already concise, helpers added no value
  - Used errors.Join in wrapError - Preserves original AWS SDK error for diagnostics while matching sentinel errors
  - Converted wrapError to package-level function

liuzengh reviewed (approved):
LGTM
